presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/BaseCostBasedPlanTest.java

```diff
@@ -0,0 +1,248 @@
+
+    @DataProvider
+    public Object[][] getQueriesDataProvider()
+    {
+        return getQueries()
+                .collect(toDataProvider());
+    }
+
+    @Test(dataProvider = "getQueriesDataProvider")
+    public void test(Query query)
+    {
+        assertEquals(generateQueryPlan(query.getQuery()), read(resolve(getExpectedJoinOrderingFile(query.getId()))));
+    }
+
+    private Path resolve(Path path)
+    {
+        // Test is run with module as current working directory. generate() is run with project's top-level directory as current working directory.
+        if (exists(path)) {
+            return path;
+        }
+        Path fromParent = Paths.get("..").resolve(path);
+        checkState(exists(fromParent), "Unable to resolve: " + path);
+        return fromParent;
+    }
+
+    public void generate()
+            throws Exception
+    {
+        initPlanTest();
+        try {
+            getQueries()
+                    .parallel()
+                    .forEach(query -> {
+                        try {
+                            Files.write(generateQueryPlan(query.getQuery()).getBytes(UTF_8), getExpectedJoinOrderingFile(query.id).toFile());
+                            System.out.println(query.getId());
+                        }
+                        catch (IOException e) {
+                            throw new UncheckedIOException(e);
+                        }
+                    });
+        }
+        finally {
+            destroyPlanTest();
+        }
+    }
+
+    private static String read(Path file)
+    {
+        try {
+            return Files.asCharSource(file.toFile(), UTF_8).read();
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    public static final class Query
+    {
+        private final String id;
+        private final String query;
+
+        public Query(String id, String query)
+        {
+            this.id = requireNonNull(id, "id is null");
+            this.query = requireNonNull(query, "query is null");
+        }
+
+        public String getId()
+        {
+            return id;
+        }
+
+        public String getQuery()
+        {
+            return query;
+        }
+
+        @Override
+        public String toString()
+        {
+            // Keep toString() short, as this is used in @DataProvider
+            return id;
+        }
+    }
+
+    private String generateQueryPlan(String query)
+    {
+        String sql = query.replaceFirst(";", "")
+                .replace("${database}.${schema}.", "")
+                .replace("\"${database}\".\"${schema}\".\"${prefix}", "\"");
+        Plan plan = plan(sql, LogicalPlanner.Stage.OPTIMIZED_AND_VALIDATED, false);
+
+        JoinOrderPrinter joinOrderPrinter = new JoinOrderPrinter();
+        plan.getRoot().accept(joinOrderPrinter, 0);
+        return joinOrderPrinter.result();
+    }
+
+    private static class JoinOrderPrinter
```

Comment:
In a follow up I would like to extract this concept and use it (replace the usage) instead of `com.facebook.presto.sql.planner.assertions.PlanMatchPattern`. This is much more easier to debug and maintain.

Reply:
cc: @atris